### PR TITLE
[core][updates] Add bridgeless support for ReactNativeHostHandler

### PR DESCRIPTION
**Pull request by Kudo**
# Why

add bridgeless support for our react delegate on android
close ENG-11646

# How

- we cannot use `getDefaultReactHost()` because we will fit the `You can call getDefaultReactHost only with instances of DefaultReactNativeHost` exception. since our reactNativeHost is actually a ReactNativeHostWrapper. this pr tries to add a `ExpoReactHostFactory` to handle that.
- refactor `ReactNativeHostHandler` a little bit to be ReactHost/ReactInstanceManager independent
- drop react-native < 0.74.0 support

# Test Plan

- ci passed
- test on fabric-tester
  - removing expo-av, since expo-av internally access to `getCatalystInstance()` that we could address later.
  - apply patches from #27565
  - apply patch
  ```patch
    diff --git a/packages/expo-modules-core/android/src/main/java/expo/modules/adapters/react/NativeModulesProxy.java b/packages/expo-modules-core/android/src/main/java/expo/modules/adapters/react/NativeModulesProxy.java
    index 0448aad361..4d52e3d062 100644
    --- a/packages/expo-modules-core/android/src/main/java/expo/modules/adapters/react/NativeModulesProxy.java
    +++ b/packages/expo-modules-core/android/src/main/java/expo/modules/adapters/react/NativeModulesProxy.java
    @@ -118,7 +118,7 @@ public class NativeModulesProxy extends ReactContextBaseJavaModule {
        * or `NativeModulesProxy.callMethod("NativeDictionary", 2, ["key", 42])`, where the second argument
        * is a method's constant key.
        */
    -  @ReactMethod
    +//  @ReactMethod
       public void callMethod(String moduleName, Dynamic methodKeyOrName, ReadableArray arguments, final Promise promise) {
         String methodName;
         if (methodKeyOrName.getType() == ReadableType.String) {
    diff --git a/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ReactExtensions.kt b/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ReactExtensions.kt
    index 6a79b5c101..1584f615d2 100644
    --- a/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ReactExtensions.kt
    +++ b/packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/ReactExtensions.kt
    @@ -1,6 +1,7 @@
     package expo.modules.kotlin
    
     import com.facebook.react.bridge.ReactContext
    +import com.facebook.react.uimanager.ThemedReactContext
     import expo.modules.adapters.react.NativeModulesProxy
    
     internal fun ReactContext.getUnimoduleProxy(): NativeModulesProxy? {
    @@ -8,6 +9,6 @@ internal fun ReactContext.getUnimoduleProxy(): NativeModulesProxy? {
       return if (!isBridgeless) {
         catalystInstance?.getNativeModule("NativeUnimoduleProxy") as? NativeModulesProxy
       } else {
    -    nativeModules?.find { it is NativeModulesProxy } as? NativeModulesProxy
    +    (this as? ThemedReactContext)?.reactApplicationContext?.nativeModules?.find { it is NativeModulesProxy } as? NativeModulesProxy
       }
     }
  ```
  - linked expo-updates and do some test (i believe that's not fully work, should follow up later)

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by Kudo** [comment]
updates e2e failed because it's running on 0.73. will have a pr to migrate all our repo and template to 0.74

**Review by wschurman — APPROVED** [review]
expo-updates changes look good. will defer to other reviewers for the other parts.

File: packages/expo/CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@
 - Use `typeof window` checks for removing server code. ([#27514](https://github.com/expo/expo/pull/27514) by [@EvanBacon](https://github.com/EvanBacon))
 - [expo-updates] Migrate to requireNativeModule/requireOptionalNativeModule. ([#25648](https://github.com/expo/expo/pull/25648) by [@wschurman](https://github.com/wschurman))
 - Remove implicit dependency on expo-updates to do runtime version check at runtime. ([#26080](https://github.com/expo/expo/pull/26080) by [@wschurman](https://github.com/wschurman))
+- [Android] Added bridgeless support on ReactNativeHostHandler and drop support. Also dropped support for React Native 0.73 and lower. ([#27629](https://github.com/expo/expo/pull/27629) by [@kudo](https://github.com/kudo))
```
Comment (tsapeta) [inline packages/expo/CHANGELOG.md:20]:
```suggestion
- [Android] Added bridgeless support on ReactNativeHostHandler and dropped support for React Native 0.73 and lower. ([#27629](https://github.com/expo/expo/pull/27629) by [@kudo](https://github.com/kudo))
```

But maybe it should be added as a separate entry in the breaking changes section?

**Review by tsapeta — APPROVED** [review]

**Review by lukmccall — APPROVED** [review]
LGTM

**Pull request merged**